src/Workspaces/Remote/ServiceHub/Services/CustomMessageHandler/RemoteCustomMessageHandlerService.cs

```diff
@@ -0,0 +1,252 @@
+    /// </summary>
+    private readonly ConcurrentDictionary<string, CustomMessageHandlerExtension> _extensions = new();
+
+    private readonly System.Runtime.Loader.AssemblyLoadContext _defaultLoadContext
+        = System.Runtime.Loader.AssemblyLoadContext.GetLoadContext(typeof(RemoteCustomMessageHandlerService).Assembly)
+        ?? throw new InvalidOperationException($"Cannot get assembly load context for {nameof(RemoteCustomMessageHandlerService)}.");
+#endif
+
+    internal sealed class Factory : FactoryBase<IRemoteCustomMessageHandlerService>
+    {
+        protected override IRemoteCustomMessageHandlerService CreateService(in ServiceConstructionArguments arguments)
+            => new RemoteCustomMessageHandlerService(arguments);
+    }
+
+    public RemoteCustomMessageHandlerService(in ServiceConstructionArguments arguments)
+        : base(arguments)
+    {
+    }
+
+    public ValueTask<HandleCustomMessageResponse> HandleCustomMessageAsync(
+        Checksum solutionChecksum,
+        string assemblyPath,
+        string typeFullName,
+        string jsonMessage,
+        DocumentId? documentId,
+        ImmutableArray<LinePosition> positions,
+        CancellationToken cancellationToken)
+    {
+#if NETSTANDARD2_0
+        throw new InvalidOperationException("Custom handlers are only supported in the servicehub host");
+#else
+
+#if DEBUG
+        System.Diagnostics.Debugger.Launch();
+#endif
+
+        Requires.NotNullOrEmpty(assemblyPath);
+        Requires.NotNullOrEmpty(typeFullName);
+        Requires.NotNullOrEmpty(jsonMessage);
+        Requires.NotNullOrEmpty(positions);
+
+        return RunServiceAsync(solutionChecksum, async solution =>
+        {
+            Document? document = null;
+            if (documentId is not null)
+            {
+                document = solution.GetDocument(documentId) ?? await solution.GetSourceGeneratedDocumentAsync(documentId, cancellationToken).ConfigureAwait(false);
+                Contract.ThrowIfNull(document);
+            }
+
+            try
+            {
+                var extension = _extensions.GetOrAdd(assemblyPath, _ =>
+                {
+                    var directory = Path.GetDirectoryName(assemblyPath);
+                    Contract.ThrowIfNull(directory);
+
+                    var loadContext = new System.Runtime.Loader.AssemblyLoadContext(name: $"RemoteCustomMessageHandlerService assembly load context for {assemblyPath}", isCollectible: true);
+                    loadContext.Resolving += (context, assemblyName) =>
+                    {
+                        var sharedAssembly = _defaultLoadContext.Assemblies.Where(a => a.GetName().Name == assemblyName.Name).FirstOrDefault();
+
+                        if (sharedAssembly is not null)
+                        {
+                            if (sharedAssembly.GetName().Version < assemblyName.Version)
```

Comment:
I don't think we should do any kind of version checks on this.  Similar to the above, we just delegate to `LoadFromAssemblyName` and catch if it throws.